docs: add the proposal for Notation blob signing and verification (#1180)
notaryproject/notation: specs/proposals/blob-signing.md

# Proposal for Supporting Blob Signing and Verification in Notation

## Overview 

> A **blob**, short for [Binary Large Object](https://wikipedia.org/wiki/Binary_large_object) is a collection of binary data stored as a single entity. In this proposal, a blob refers to any binary data or file, including SBOM (Software Bill of Materials) files, release assets, AI model files, WebAssembly files or other forms of unstructured data. While Notation currently supports signing and verifying OCI (Open Container Initiative) artifacts such as container images, this document outlines a proposal to extend the Notation tool's capabilities to include blob signing and verification.

Reviewer: Just some notes for AI models:
- **Ollama** --> OCI-like Artifacts
- **LM Studio** --> Hugging Face (individual files)

Author: Thanks.

## Problem Statement & Motivation 

Signing container images alone is not enough to secure the software supply chain for containers. In addtion to images, SBOMs, configuration files, release binaries or other artifacts must also be signed and verified to ensure their integrity and authenticity. With the increasing use of AI model files and WebAssembly files, securing these artifacts is equally critical. By adding blob signing, Notation can extend its capabilities to handle a broader range of artifacts across various scenarios, enhancing its versatility while leveraging its existing key management and trust policy framework.

## Scenarios 

The following describes how blob signing and verification can be used for various scenarios:

### Scenario 1: Blob signing and verification with file-based distribution

> Sarah, a software engineer, creates binaries for applications that are distributed via her company's website. In CI/CD pipelines, she creates tasks that generates an SBOM for each binary and signs the SBOM to ensure the authenticity and integrity of the SBOM. The pipeline automatically generates a digital signature for each SBOM, which is stored alongside the SBOM on the filesystem. With SBOM and signatures created, the pipeline tasks automatically upload the binaries, SBOMs, and SBOM signatures to the company website as separate downloadable artifacts. The compliance team downloads the SBOM and signature to check whether the corresponding binary is compliant with both company and governance rules. Before analyzing the SBOM, they verify its integrity against the corresponding signature, ensuring the SBOM hasn't been tampered with since being signed in the pipeline and that it originates from the Sarah's company.

Reviewer: nit: just curious why Sarah :D

Usually,
- A --> Alice
- B --> Bob
- C --> Charlie
- D --> David

Basically, those names are used to replace User A, User B, User C, bla bla...

Author: It was generated by AI :)

## Proposal 

This section outlines the proposed solution for using Notation CLI commands to sign and verify blobs. The following points are not covered in this document:
- Detailed command usage, which will be covered in the individual command specifications at [Notation CLI Docs](https://github.com/notaryproject/notation/tree/v1.3.0/specs/commandline).
- Distribution of certificates for verification across different scenarios.

> - Compatibility of the payload format. While some users are utilizing Notary Project-compliant signature envelopes, the payload format for blobs differs from the [Notary Project specification](https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md#blob-payload). 

Author: Discussed with Toddy, it is mainly about payload format of blob signature is different from other tools which generates a different payload format, for example, only digest is included as the payload format. I will start a new discusssion thread about the payload format, the goal is no matter which tool produces Notary Project signatures, it can be processed by Notation and other tools per the Notary Project signature specification. /cc: @shizhMSFT @Two-Hearts @JeyJeyGao

- CI/CD experience. This document specifically focuses on using Notation CLI commands.

In general, all blob-related commands are grouped under the `notation blob` command for easy discovery and to avoid mixing them with OCI-based artifact commands.

### Blob Sign and verify with file-based distribution

For file-based distribution, such as SBOMs or release artifacts shared via a website or file transfer system, the process of signing and verifying blobs follows a similar flow. The steps below outline the user experience for signing and verifying a file, using an SBOM file named `sbom.json` as an example.

**Prerequisites:**
- SBOMs, release artifacts or other files are created and ready for publishing

**Steps:**

1. Sign a file on filesystem with a specified signature envelope:

    ```shell
    notation blob sign --signature-format cose sbom.json
    ```

>     This generates a signature named `sbom.json.cose.sig` in the same directory as `sbom.json`. The signature file name follows the [Notary Project specification](https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md#blob-signatures).

Author: No, we will not change the current behavior. I will update the text.

>     This generates a signature named `sbom.json.cose.sig` in the same directory as `sbom.json`. The signature file name follows the [Notary Project specification](https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md#blob-signatures).

Reviewer 2: > Can we have default option and overwrite? For example, somebody may want to store in specific folder.

Yes, we do have this overwrite option. It's just the default location that needs to be discussed here.

>     This generates a signature named `sbom.json.cose.sig` in the same directory as `sbom.json`. The signature file name follows the [Notary Project specification](https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md#blob-signatures).

Reviewer: Please use a versioned link or a link with commit hash.

Author: It seems the blob spec was not released, so it cannot use versioned link.

Author: The blob spec was not released yet, so I cannot use a versioned link, will use a commit hash

>     This generates a signature named `sbom.json.cose.sig` in the same directory as `sbom.json`. The signature file name follows the [Notary Project specification](https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md#blob-signatures).

Author: @toddysm This document provides an overview of commands, while additional flags will be detailed in the `notation blob sign` command specification. Each command's specification will include complete flags and usages.

2. Publish the file and corresponding signature using any file transfering mechanism. Both the file and its signature can be packaged into one file (e.g., a tarball) for transferring.

3. Download or fetch both the file and signature.

4. Verify the file against the signature:

   - Add the root CA certificate and set up the trust store (The same experience for verifying container images):

        ```shell
        notation cert add --type ca --store myCACerts root.crt
        ```

     Confirm the certificate is added.

        ```shell
        notation cert ls
        ```

>         > NOTE
>         >
>         > Learn more options using `notation cert ls --help`.

Reviewer 2: nit:
```suggestion
        > [!NOTE]
        > Learn more options using `notation cert ls --help`.
```

    - Set up trust policy for blobs by adding a new command `notation blob policy init`. This command streamlines the process, eliminating the need for users to consult documentation for the correct trust policy format and preventing the accidental use of policies intended for other verification purposes.

        ```shell

>         notation blob policy init --name "myBlobPolicy" --trust-store "ca:myCACerts" --trust-identity "x509.subject:C=US,ST=WA,O=wabbit-network.io"

Reviewer 3: I think the verification level is `strict` and GlobalPolicy is true when using blob policy init command, right?

Author: Yes, other parameters will use the default values. Users can update them later according to the update guideline.

Reviewer 4: How is the blob policy different from the standard Notation policy? Why do we need to have commands in `notation blob policy` and `notation policy`?

Author: @toddysm It is according to the spec https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#trust-policy-properties, which has been discussed and aligned within the community when AWS folks worked on this feature. Blob will use a slight different trust policy format in a different trust policy config file. So to make the experience simple for users, the new command `init` will help users to intialize a trust policy easily with default secure settings. Later on, it can be expanded to `notation policy init` command for OCI artifact, so that users do not need to do any manual edit for default config.

Author: See the updated comment at https://github.com/notaryproject/notation/pull/1180#discussion_r1966433336

Author: Aligend after offline discussion. We will follow the existing [blob CLI spec](https://github.com/notaryproject/notation/blob/main/specs/commandline/blob.md) for implementing the commands.

        ```

        > NOTE
        >
        > See the [section](#manage-blob-policies) for more commands for `notation blob policy`.
        >

>         > This `init` command can be extended to policy commands for OCI artifacts.

Author: I don't have strong opinion. I can remove it.

    - Verify the signature:

        ```shell
        notation blob verify --policy-name myBlobPolicy --signature sbom.json.cose.sig sbom.json
        ```

        If a [global policy](https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#blob-trust-policy) is set, you can skip the `--policy-name` flag.

> ### Blob Sign and verify with registry-based distribution

Reviewer: I have a fundamental question: What the difference and advantage of "Blob Sign and verify with registry-based distribution", comparing with the following workflow?
```sh
# sign
notation blob sign --signature-format cose sbom.json
oras push myregistry/mypath/mysboms:v1 sbom.json sbom.json.cose.sig

# verify
oras pull myregistry/mypath/mysboms:v1
notation blob verify --policy-name myBlobPolicy --signature sbom.json.cose.sig sbom.json
```
Basically, pack the sbom with its signature as a single OCI artifact

Why I'm asking since there will only be one signature for a blob for the most cases. Registries are one of mediums just like the file system.

Reviewer: The only advantage of the registries is that the registry may have the digest pre-computed in the manifest. However, the digest algorithm may not match the digest algorithm in the signature payload. /cc @Two-Hearts

Reviewer 2: This is a question on the use case.

@yizha1 Do we want to support the following scenario:
Suppose I have a file signed in my local filesystem, and I used oras to push both the file and signature to the registry. For some reason, I want to re-sign the file. What I need to do is `notation blob sign` -> `oras push` (since the file and the first signature was packed as a OCI image instead of a referrer, I need a second image to verify the second signature). The verifier would then need to do `oras pull` all applicable signatures-> `notation blob verify` for each pulled signature.
If we have the design suggested by @yizha1, then the workflow becomes:
`notation blob sign` again, and then `notation blob push`. On verification side, `notation blob verify`, Notation would verify every applicable signature on the fly.

If we only want to support one signature per blob file at any time point, then yes, both designs would do the job. But if as the time goes by, the number of signatures could increase, then the registry way is better. /cc: @shizhMSFT

Reviewer 2: And a potential security question regarding `oras pull` both the blob and the signature to local file system for verification:
Based on Notation's [threat model](https://github.com/notaryproject/specifications/blob/main/threatmodels/notation-threatmodel.md#notation-verify-artifacts-against-signatures), the local file system is within the trust boundary, but remote registry is not. Downloading the blob from a registry to local file system before doing any Notation verification can be risky?

Author: Discussed with @Two-Hearts, the threat model will be tracked by the existing issue https://github.com/notaryproject/specifications/issues/296

For registry-based distribution, such as using an OCI-compliant container registry, the process is similar but includes additional steps for pushing and verifying blobs in the registry.

**Prerequisites:**
- SBOMs, release artifacts or other files are created and ready for publishing.

**Steps:**

1. Sign a file on filesystem with a specified signature envelope:

    ```shell
    notation blob sign sbom.json --signature-format cose --id myKeyId --plugin myKMSPluginName
    ```

    This generates a signature named `sbom.json.cose.sig` in the same directory as `sbom.json`.

2. Publish the file and signature to the OCI-compliant registry:

    ```shell
    notation blob push --artifact-type "application/spdx+json" sbom.json:"application/json" --signature sbom.json.cose.sig --reference myregistry/mypath/mysboms:v1
    ```

>     This command stores `sbom.json` as an OCI image referenced by `myregistry/mypath/mysboms:v1` in the registry, which has the artifact type `application/spdx+json` and `sbom.json` data stored as a layer with the media type `application/json`. Additionally, this command stores the signature file `sbom.json.cose.sig` as another OCI image with the artifact type `application/vnd.cncf.notary.signature` and the signature data as a layer with media type `application/cose`. The artifact type should be `application/vnd.cncf.notary.signature` as it is Notary Project signature. The layer media type can be detected according to the file name, which is either `application/cose` or `application/jws`. The signature manifest's `subject` property refers to the `myregistry/mypath/mysboms:v1`. To distinguish these signatures from those generated during OCI artifact signing, the specific annotation `io.cncf.notary.blob.signature=true` is added for the signature manifest.

Author: We can discuss it. From users point of view, this will provide better experience. If users only want to push the blob, we should not support, but users want to push both blob and signature, this will make it easy for users.

>     This command stores `sbom.json` as an OCI image referenced by `myregistry/mypath/mysboms:v1` in the registry, which has the artifact type `application/spdx+json` and `sbom.json` data stored as a layer with the media type `application/json`. Additionally, this command stores the signature file `sbom.json.cose.sig` as another OCI image with the artifact type `application/vnd.cncf.notary.signature` and the signature data as a layer with media type `application/cose`. The artifact type should be `application/vnd.cncf.notary.signature` as it is Notary Project signature. The layer media type can be detected according to the file name, which is either `application/cose` or `application/jws`. The signature manifest's `subject` property refers to the `myregistry/mypath/mysboms:v1`. To distinguish these signatures from those generated during OCI artifact signing, the specific annotation `io.cncf.notary.blob.signature=true` is added for the signature manifest.

Reviewer 4: > This command stores `sbom.json` as an OCI image...

Do you mean _OCI image_ or _OCI artifact_?

Author: It is an OCI artifact. I want to say they are stored per the OCI image spec. I will rephase it.

>     This command stores `sbom.json` as an OCI image referenced by `myregistry/mypath/mysboms:v1` in the registry, which has the artifact type `application/spdx+json` and `sbom.json` data stored as a layer with the media type `application/json`. Additionally, this command stores the signature file `sbom.json.cose.sig` as another OCI image with the artifact type `application/vnd.cncf.notary.signature` and the signature data as a layer with media type `application/cose`. The artifact type should be `application/vnd.cncf.notary.signature` as it is Notary Project signature. The layer media type can be detected according to the file name, which is either `application/cose` or `application/jws`. The signature manifest's `subject` property refers to the `myregistry/mypath/mysboms:v1`. To distinguish these signatures from those generated during OCI artifact signing, the specific annotation `io.cncf.notary.blob.signature=true` is added for the signature manifest.

Reviewer 2: @shizhMSFT @FeynmanZhou @JeyJeyGao Looking for your insights here.

>     This command stores `sbom.json` as an OCI image referenced by `myregistry/mypath/mysboms:v1` in the registry, which has the artifact type `application/spdx+json` and `sbom.json` data stored as a layer with the media type `application/json`. Additionally, this command stores the signature file `sbom.json.cose.sig` as another OCI image with the artifact type `application/vnd.cncf.notary.signature` and the signature data as a layer with media type `application/cose`. The artifact type should be `application/vnd.cncf.notary.signature` as it is Notary Project signature. The layer media type can be detected according to the file name, which is either `application/cose` or `application/jws`. The signature manifest's `subject` property refers to the `myregistry/mypath/mysboms:v1`. To distinguish these signatures from those generated during OCI artifact signing, the specific annotation `io.cncf.notary.blob.signature=true` is added for the signature manifest.

Reviewer: I agree with @Two-Hearts

>     This command stores `sbom.json` as an OCI image referenced by `myregistry/mypath/mysboms:v1` in the registry, which has the artifact type `application/spdx+json` and `sbom.json` data stored as a layer with the media type `application/json`. Additionally, this command stores the signature file `sbom.json.cose.sig` as another OCI image with the artifact type `application/vnd.cncf.notary.signature` and the signature data as a layer with media type `application/cose`. The artifact type should be `application/vnd.cncf.notary.signature` as it is Notary Project signature. The layer media type can be detected according to the file name, which is either `application/cose` or `application/jws`. The signature manifest's `subject` property refers to the `myregistry/mypath/mysboms:v1`. To distinguish these signatures from those generated during OCI artifact signing, the specific annotation `io.cncf.notary.blob.signature=true` is added for the signature manifest.

Author: Agree. I will keep only pushing signatures

>     This command stores `sbom.json` as an OCI image referenced by `myregistry/mypath/mysboms:v1` in the registry, which has the artifact type `application/spdx+json` and `sbom.json` data stored as a layer with the media type `application/json`. Additionally, this command stores the signature file `sbom.json.cose.sig` as another OCI image with the artifact type `application/vnd.cncf.notary.signature` and the signature data as a layer with media type `application/cose`. The artifact type should be `application/vnd.cncf.notary.signature` as it is Notary Project signature. The layer media type can be detected according to the file name, which is either `application/cose` or `application/jws`. The signature manifest's `subject` property refers to the `myregistry/mypath/mysboms:v1`. To distinguish these signatures from those generated during OCI artifact signing, the specific annotation `io.cncf.notary.blob.signature=true` is added for the signature manifest.

Reviewer 5: Agree with @Two-Hearts  . Keeping the scope of notation focus on signature generation and managment makes more sense.

If there is a blob file has a signature and other attestations,  should notation supports pushing blob, its signature, and other attestations together? That may beyond the scope.

>     This command stores `sbom.json` as an OCI image referenced by `myregistry/mypath/mysboms:v1` in the registry, which has the artifact type `application/spdx+json` and `sbom.json` data stored as a layer with the media type `application/json`. Additionally, this command stores the signature file `sbom.json.cose.sig` as another OCI image with the artifact type `application/vnd.cncf.notary.signature` and the signature data as a layer with media type `application/cose`. The artifact type should be `application/vnd.cncf.notary.signature` as it is Notary Project signature. The layer media type can be detected according to the file name, which is either `application/cose` or `application/jws`. The signature manifest's `subject` property refers to the `myregistry/mypath/mysboms:v1`. To distinguish these signatures from those generated during OCI artifact signing, the specific annotation `io.cncf.notary.blob.signature=true` is added for the signature manifest.

Author: Aligned offline that users will use oras to push blob file and blob signatures, and berfore verification, users need to download the blob file and signatures to filesystem and then use `notation blob verify` command for verification.

        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "artifactType": "application/vnd.cncf.notary.signature",
        "config": {
            "mediaType": "application/vnd.oci.empty.v1+json",
            "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
            "size": 2,
            "data": "e30="
        },
        "layers": [
            {
            "mediaType": "application/cose",
            "digest": "sha256:6dc2a336b124d625f87deb3f539fa1465414ddad300a898adf4682e530d6f144",
            "size": 1479,
            "annotations": {
                "org.opencontainers.image.title": "sbom.json.cose.sig"
            }
            }
        ],
        "subject": {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:57eed31a09556933bddd4e164e1e085a220cf00489477eb98d89feac902eb42b",
            "size": 553
        },
        "annotations": {
            "io.cncf.notary.blob.signature": "true",
            "io.cncf.notary.x509chain.thumbprint#S256": "[\"aaaa\", \"bbbb\"]",
            "org.opencontainers.image.created": "2025-02-18T06:08:45Z"
        }
    }
    ```

>     Alternatively, you can also use ORAS tool to push the blob file first, and then use Notation CLI command to push only the signature:

Reviewer 2: ```suggestion
    Alternatively, you can also use other tools such as ORAS to push the blob file first, and then use Notation CLI command to push the signature:
```

>     Alternatively, you can also use ORAS tool to push the blob file first, and then use Notation CLI command to push only the signature:

Reviewer 4: As this only requires pushing artifacts to the registry, can we have both, the SBOM and the signature, pushed by ORAS? From experience point of view, mixing the tools degrades the experience. So, thinking about the experience, Notation is for "signing", so, it is used to sign the artifact (SBOM). ORAS is for pushing and pulling artifacts, so, it is used to push both, the SBOM and the signaturel.

Author: Manage signatures is with Notation's scope. So, considering the comments from Patrick and Shiwei, I will remove the function of using Notation to push both blob and signature, but keep `notation blob push` for pushing signatures. If users want to use oras to push signature as referrers, then the command looks like below:

```
oras attach <my_registry>/<my_repo>:v1 --artifact-type 'application/vnd.cncf.notary.signature' myFile.cose.sig:application/cose -a "io.cncf.notary.blob.signature=true" -a "io.cncf.notary.x509chain.thumbprint#S256=[\"aaaa\", \"bbbb\"]"
```

As you can see, it is not a good experience for users, epecially the thumbprint parameter. 

Shiwei's has a good comment, https://github.com/notaryproject/notation/pull/1180#discussion_r1961950312, if we agree with this, then we probably do not need `notation blob push`.

Author: Aligend on the comment https://github.com/notaryproject/notation/pull/1180#discussion_r1976467426

    ```shell
    oras push myregistry/mypath/mysboms:v1 sbom.json:application/octet-stream 

>     notation blob push --signature sbom.json.cose.sig --reference myregistry/mypath/mysboms:v1

Author: `notation blob push` command will not be introduced based on alignment. See https://github.com/notaryproject/notation/pull/1180#discussion_r1976467426

    ```

>     You can also create an oci image layout and then use ORAS tool to copy it to the target registry. 

Reviewer 2: Since oci image layout is an experimental feature in Notation, I think we should remove it from this doc for now.

Author: I would like to keep it to avoid missing this case in the future.

>     You can also create an oci image layout and then use ORAS tool to copy it to the target registry. 

Reviewer 4: Why does it need to be OCI layout?

Author: This is another opition. That users can create OCI layout for signature first and then copy to the registry.

Author: Remove OCI image layout from the proposal. It will be brought back if there is a clear scenario.

    ```shell
    notation blob push --artifact-type "application/spdx+json" sbom.json:"application/json" --signature sbom.json.cose.sig --oci-layout sbom.json:v1
    oras copy -r --from-oci-layout sbom.json:v1 myregistry/mypath/mysboms:v1
    ```

    > NOTE:
    >
    > Support for OCI image layout in Notation is currently an experimental feature.

3. Verify the signature from the registry:

    - Add the root CA certificate and set up the trust store:

        ```shell
        notation cert add --type ca --store myCACerts root.crt
        ```

        Confirm that the cert is added.

        ```shell
        notation cert ls
        ```

    - Set up the trust policy:

        ```shell

>         notation blob policy init --name "myBlobPolicy" --trust-store "ca:myCACerts" --trust-identity "x509.subject:C=US,ST=WA,O=wabbit-network.io"

Reviewer 4: Again, why do we need to have `policy` command and `blob policy` subcommand? This is suboptimal experience IMHO

Author: See this comment https://github.com/notaryproject/notation/pull/1180#discussion_r1962855581

Author: Toddy and I discussed offline and agreed that the current proposal for managing trust policies for OCI artifacts and blobs is not optimal.  We should explore a better solution due to potential user confusion with the two policy commands and configuration files. /cc @shizhMSFT @Two-Hearts @JeyJeyGao

We also need to consider the release version, if we starts with v2, we may break the experience later on and require a major version stepping up. One option is that we could do v1.4.0 release first, and improve the experience and other command experience in v2 including trust policy format.

        ```

    - Verify the signature:

        ```shell

>         notation blob verify --reference myregistry/mypath/mysboms:v1 --policy-name myBlobPolicy

Reviewer 5: If the blob and its signature have been pushed to the registry, how can an end-user differentiate between a blob and a normal OCI image in the registry? It determines which `notation` command to verify.

Author: The blob signature manifest will have an annotation `io.cncf.notary.blob.signature: "true"`.

        ```

### Manage blob policies

The following commands are available for managing blob policies:

- Initialize blob policies:

    ```shell

>     notation blob policy init --name "myBlobPolicy" --trust-store "ca:myCACerts" --trust-identity "x509.subject:C=US,ST=WA,O=wabbit-network.io"

Author: Yes, but users need to specify it clearly, we will not by defaut use `*`, which could be security risk. I will add a note in the document

# Proposal for Supporting Blob Signing and Verification in Notation

## Overview 

> A **blob**, short for [Binary Large Object](https://wikipedia.org/wiki/Binary_large_object) is a collection of binary data stored as a single entity. In this proposal, a blob refers to any binary data or file, including SBOM (Software Bill of Materials) files, release assets, AI model files (such as model files on [Hugging Face](https://huggingface.co/)), WebAssembly files or other forms of unstructured data. While Notation currently supports signing and verifying OCI (Open Container Initiative) artifacts such as container images, this document outlines a proposal to extend the Notation tool's capabilities to include blob signing and verification.

Reviewer: Note: Technically, a file and a blob are different. A file has a header with create time, mod time, user, group, permission, etc. while a blob has only the content data.

## Proposal 

This section outlines the proposed solution for signing and verifying blobs using Notation CLI commands. The following topics are outside the scope of this document:  

- **Detailed command usage**, which is covered in the individual command specifications at [Notation CLI specs](https://github.com/notaryproject/notation/tree/main/specs/commandline).  
- **Distribution of certificates** for verification in different scenarios.  
- **Blob signature definitions**, as outlined in the [Notary Project specification](https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md#blob-payload).  
- **CI/CD integration**, as this document specifically focuses on using Notation CLI commands.  

In general, all blob-related commands are grouped under the `notation blob` command per the [blob command spec](https://github.com/notaryproject/notation/blob/main/specs/commandline/blob.md).

> ### Blob Sign and verify with file-based distribution

Reviewer: nit:
```suggestion
### Blob sign and verify with file-based distribution
```

2. Publish both the file and corresponding signature using any file transfering mechanism. Both the file and its signature can be packaged into one file (e.g., a tarball) for transferring.

3. Download or fetch both the file and signature. If they are packaged into one file, for example, a tarball, unpackage the file to get separate file and signature.

4. Verify the file against the signature:

   - Add the root CA certificate and set up the trust store (The same experience for verifying container images):

        ```shell
        notation cert add --type ca --store myCACerts root.crt
        ```

     Confirm the certificate is added.

        ```shell
        notation cert ls
        ```

        > [!NOTE]

>         > Learn more options using `notation cert ls --help`.

Reviewer: nit: It seems those two lines are not rendered well on GitHub in a list. We probably need to reformat it for better rendering.

    - Verify the signature:

        ```shell
        notation blob verify --policy-name myBlobPolicy --signature sbom.json.cose.sig sbom.json
        ```

        If a [global policy](https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#blob-trust-policy) is set, you can skip the `--policy-name` flag.

### Blob Sign and verify with registry-based distribution

For registry-based distribution, such as using an OCI-compliant container registry, the process is similar but includes additional steps for pushing blobs and sigantures to the registry.

**Prerequisites:**
- SBOMs, release artifacts or other files are created and ready for publishing.

**Steps:**

1. Sign a file on filesystem with a specified signature envelope:

    ```shell
    notation blob sign --id myKeyId --signature-format "cose" --media-type "application/spdx+json" sbom.json
    ```

    This generates a signature named `sbom.json.cose.sig` in the current working directory.

2. Publish the file and signature to the OCI-compliant registry:

    Use [ORAS tool](https://oras.land/) to push the blob file:

    ```shell

>     oras push --artifact-type "application/spdx+json" sbom.json:application/json myregistry/myapp/sbom:v1

Reviewer: Can I also use `oras attach`?

Author: This command pushes the SBOM file, while `oras attach` is used later to attach the signature.

In this scenario, the SBOM is generated for a software application, not for a container image. Therefore, the example uses `oras push`, not to attach the SBOM to a container image in the registry.
